libff/algebra/curves/tests/test_groups.cpp

```diff
@@ -317,6 +317,26 @@ void test_bls12_377()
     ASSERT_EQ(bls12_377_G2::zero(), z);
 }
 
+// chack that some elements e.g. 1,-1,2,random satisfy the curve
+// equation Y^2 = X^3 + a X + b; used in test_bls12_381
+template<typename GroupT> void check_curve_equation(GroupT P)
+{
+    P.to_affine_coordinates();
+    using Fq = typename std::decay<decltype(P.X)>::type;
+    Fq LHS = (P.Y * P.Y);
+    Fq RHS = ((P.X * P.X * P.X) + (GroupT::coeff_a * P.X) + GroupT::coeff_b);
+    ASSERT_EQ(LHS, RHS);
+}
+
+template<typename GroupT> void test_bls12_381()
+{
+    using Fr = typename GroupT::scalar_field;
+    check_curve_equation(GroupT::one());
+    check_curve_equation(Fr(-1l) * GroupT::one());
+    check_curve_equation(Fr(2l) * GroupT::one());
```

Comment:
Any specific rationale behind the 2 test cases above?

Reply:
What about negative test cases?

Reply:
No specific rationale TBH. Just following the approach from https://github.com/clearmatics/libff/blob/develop/libff/algebra/curves/tests/test_groups.cpp#L59 with the goal to catch some immediate errors in the curve implementation.

By negative test case do you mean a test case that is expected to fail? Let's discuss this when we meet.

Reply:
Hi, @AntoineRondelet! For the negative test case, did you have in mind to do it both for the `G1` and `G2` curves? I am asking because the `G2` case looks slightly more complicated. The elements of the `G2` curve are in the twist field `bls12_381_Fq2` and as such are represented by coefficients `c0,c1`:

https://github.com/clearmatics/libff/blob/bls12-381/libff/algebra/fields/fp2.hpp#L66-L71

https://github.com/clearmatics/libff/blob/bls12-381/libff/algebra/curves/bls12_381/bls12_381_g2.hpp#L49-L51

I am currently reading up on twist fields and their representation (and the meaning of the coefficients `c0,c1`, in particular) but I would appreciate some help with initializing an (invalid) point on the `G2` curve via the constructor.

Reply:
If I understood correctly, the point of this test is to check that the points generated by the group class are consistent with the a and b coefficients.  Namely it checks that things are initialized correctly.  Since `check_curve_equation` is part of the test code, we may not need negative examples in this case, but tests for methods such as `is_well_formed` should check the negative case (although we don't seem to have such tests at the moment).

However, the function `template<typename GroupT> void test_bls12_381()` is actually generic, so could be called something like `test_curve_equation` and it woudl be good to make this check for all curves.

(I think testing `is_well_formed` can be pushed to an issue for now.)

Reply:
Generally I think we need to have positive and negative test cases to properly test the code. In the trivial case where code is tautological and only positive test cases are implemented, the boundaries of the code aren't properly checked, and the test doesn't do anything really useful. So what I'm saying here in essence is: I'd had a negative test case to the `check_curve_equation` to properly check this piece of code, else:
- We generate a point on the curve, and
- Test it is on the curve

That's ok, but ideally, we should make sure that only the right subset of field element tuples pass this test, so to chck this we need to run test on invalid data and make sure it fails as expected.
Btw, this is a general point about testing, we should properly check the boundaries of the code, and to do this, we generally need more than "positive" test-cases.

Reply:
> Generally I think we need to have positive and negative test cases to properly test the code. 

Definitely in alignment here about the general testing approach.  Hence my point about the `is_well_formed` methods needing exactly both of these cases.

If I've understood this particular case (which I possibly didn't) we have a function *as part of the test* which checks that the curve eqution is satisfied for a given point, and we use this to check that some fundamental points from G1 and G2 satisfy the curve equation.  In terms of testing the negative here, the only option I see (and this is also where I may be missing something) would be to intentionally construct a curve point that does not sataisfy the curve equation, and then check that it does not satisfy the curve equation.  All within the test.  This seems to me to be a tautology, and would not involve any code outside of the test.

Let me know if that doesn't make sense, though.  If there is a negative to test I'm certainly not against adding that.

Reply:
>  In terms of testing the negative here, the only option I see (and this is also where I may be missing something) would be to intentionally construct a curve point that does not sataisfy the curve equation, and then check that it does not satisfy the curve equation. All within the test. This seems to me to be a tautology, and would not involve any code outside of the test.

Ah I see what you mean @dtebbs - sorry i didn't click on the first read. Yes, that's right, I didn't click that the `check_curve_equation` function was only a *utils _test_ function*, so yes, we should both positive and negative test cases for the equivalent _non-test_ function , which seems to be `is_well_formed` here.

Reply:
Yes, should we make an issue and discuss a few related things:
- tests for the `is_well_formed` method  (generate negative tests cases by just addign 1 to the Y coord, etc)
- use `is_well_formed` to check curve points on read?  (since it will generally be faster)  (remembering that we should also have a flag for this)
- a generic equation-checking function in the main body? (in fact some implementations of `is_well_formed` may be able to reuse it)

Reply:
Yes let's do this. Let's discuss this on a separate ticket, and get this PR merged.

Reply:
Opened ticket: https://github.com/clearmatics/libff/issues/76 to discuss all this.